### PR TITLE
grafana/ui: Update React Hook Form to v7

### LINKED ISSUE #32828: Grafana-UI: Update React Hook Form to v7
**What would you like to be added**:
Current version of the RHF in use in Grafana-UI is 5.1.3 and the latest available version is 7.0.4. We're two major versions behind and there have been quite a few changes and improvements in the RHF APIs and enhanced type support. The migration guides can be found here: 
- https://react-hook-form.com/migrate-v5-to-v6/
- https://react-hook-form.com/migrate-v6-to-v7

Since there are a lot of breaking changes, it would be a good idea for this upgrade to be a part Grafana 8. 

Also it would be a good time to consider a few issues:
- Should we move away from the `Form` component and use RHF directly in Grafana? Using this Form component has a few considerable limitations and it doesn't provide any UI value, allowing RHF to be used directly in Grafana without a wrapper would simplify a lot of things, e.g. we won't need to have a separate `FieldArray` component and `useForm` could be used at a component's top level, instead of using it's values in the render. 
- Should `InputControl` component be removed and instead RHF `Controller` used directly? `InputControl` was supposed to be an API abstraction for Control but the underlying Control's API has changed so much that it'll be easier just to use it directly. 

**Why is this needed**:
To bring up the RHF library up to date and be able to use the latest enhancements and TS support.

### PR DESCRIPTION
**Pull request by Clarity-89**
**What this PR does / why we need it**:
Update React Hook Form to the latest version. 

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana/issues/32828

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by Clarity-89** [comment]
@domasx2 could you double check that this update didn't break anything in the alerting? Didn't know we're using RHF there so extensively and it seems like a lot of the new components don't have tests there, so not sure if anything went wrong after the update. I did test the basic functionality manually and it seemed ok. Also had to remove a few types passed to the form hooks as they didn't work properly.

**Comment by domasx2** [comment]
> @domasx2 could you double check that this update didn't break anything in the alerting? Didn't know we're using RHF there so extensively and it seems like a lot of the new components don't have tests there, so not sure if anything went wrong after the update. I did test the basic functionality manually and it seemed ok. Also had to remove a few types passed to the form hooks as they didn't work properly.

Super, I'll test first thing tomorrow

File: packages/grafana-ui/src/components/Forms/FieldArray.story.tsx [hunk 1]
```diff
@@ -30,8 +30,16 @@ export const simple = () => {
                 <div style={{ marginBottom: '1rem' }}>
                   {fields.map((field, index) => (
                     <HorizontalGroup key={field.id}>
-                      <Input ref={register()} name={`people[${index}].firstName`} value={field.firstName} />
-                      <Input ref={register()} name={`people[${index}].lastName`} value={field.lastName} />
+                      <Input
+                        key={field.id}
+                        {...register(`people.${index}.firstName` as const)}
```
Comment (dprokop) [inline packages/grafana-ui/src/components/Forms/FieldArray.story.tsx:35]: `as const`? never head of such a cast.  Can you say some more about this?

File: packages/grafana-ui/src/components/Forms/Form.story.tsx [hunk 2]
```diff
@@ -74,34 +73,38 @@ const renderForm = (defaultValues?: Partial<FormDTO>) => (
...
+            <Input {...register('email', { required: true })} id="email" placeholder="roger.waters@grafana.com" />
           </Field>
 
           <Field label="Username">
-            <Input name="username" placeholder="mr.waters" ref={register} />
+            <Input {...register('username')} placeholder="mr.waters" />
           </Field>
           <Field label="Nested object">
-            <Input name="nested.path" placeholder="Nested path" ref={register} />
+            <Input {...register('nested.path')} placeholder="Nested path" />
```
Comment (dprokop) [inline packages/grafana-ui/src/components/Forms/Form.story.tsx:87]: out of curiosity, is the register path argument typed?

**Review by dprokop — APPROVED** [review]
The code looks good, haven't tested this yet tho.

One thing for the future in such migrations would be to highlight in the description how does the version bump affects our APIs.

File: packages/grafana-ui/src/components/Forms/FieldArray.story.tsx (hunk 1 above)
Comment (Clarity-89) [reply packages/grafana-ui/src/components/Forms/FieldArray.story.tsx:35]: This is necessary to strictly type check `name` and comes from TS type inference: https://devblogs.microsoft.com/typescript/announcing-typescript-4-2/#reverting-template-literal-inference

File: packages/grafana-ui/src/components/Forms/Form.story.tsx (hunk 2 above)
Comment (Clarity-89) [reply packages/grafana-ui/src/components/Forms/Form.story.tsx:87]: Afaik the types come from the default values passed to the form or fields.

**Comment by domasx2** [comment]
I fixed a small issue with rule editor form for the unified alerting, but am struggling to get receiver form to work.

It's more complicated, with nested arrays of subforms and conditional fields. And It's just breaking in frustrating ways - validating fields that were removed, losing values when new fields are added or removed and so on. It seems `react-hook-form` is imperatively maintaining a compelx internal state that is a patchwork of corner cases and workarounds... 
Considering dropping `react-hook-form` in favor of something more robust for this form </frustration>

**Comment by domasx2** [comment]
> I fixed a small issue with rule editor form for the unified alerting, but am struggling to get receiver form to work.
> 
> It's more complicated, with nested arrays of subforms and conditional fields. And It's just breaking in frustrating ways - validating fields that were removed, losing values when new fields are added or removed and so on. It seems `react-hook-form` is imperatively maintaining a compelx internal state that is a patchwork of corner cases and workarounds...
> Considering dropping `react-hook-form` in favor of something more robust for this form

nvm, made some progress

**Comment by domasx2** [comment]
Fixed some issues with receiver form, should be good now. Would be great to merge it soon then to be a able to update other PRs in flight

**Review by domasx2 — APPROVED** [review]

**Comment by Clarity-89** [comment]
> One thing for the future in such migrations would be to highlight in the description how does the version bump affects our APIs.

I have added a description of the most important breaking changes and links to the migration guides to our docs.

**Pull request merged**